nupic/research/frameworks/dynamic_sparse/networks/hebbian_v0.py

```diff
@@ -0,0 +1,252 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2019, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import torch
+from torch import nn
+
+from nupic.torch.modules import Flatten, KWinners, KWinners2d
+
+
+class HebbianNetwork(nn.Module):
+    """Parent class with shared methods. Not to be instantiated"""
+
+    def __init__(self, config=None):
+        super().__init__()
+
+    def _has_activation(self, idx, layer):
+        return (
+            idx == len(self.classifier) - 1
+            or isinstance(layer, nn.ReLU)
+            or isinstance(layer, KWinners)
+        )
+
+    def _has_params(self, layer):
+        return isinstance(layer, nn.Linear) or isinstance(layer, nn.Conv2d)
+
+    def forward(self, x):
+        if "features" in self._modules:
+            x = self.features(x)
+        x = self.classifier(x)
+        return x
+
+    def forward_with_coactivations(self, x):
+        """A faster and approximate way to track correlations"""
+
+        if "features" in self._modules:
+            x = self.features(x)
+
+        prev_act = (x > 0).detach().float()
+        for idx_layer, layer in enumerate(self.classifier):
+            # do the forward calculation normally
+            x = layer(x)
+            if self._has_params(layer) and hasattr(layer, "coactivations"):
```

Comment:
As far as I can tell, it doesn't seem as if you're tracking the coactivations: `hasattr(layer, "coactivations")` will always come up false as you never define them for a given module.

This is only a superficial observation - I haven't dived deep in the code. I may be wrong, but I just want to make sure you're certain that you're tracking coactivation as you intend.

Reply:
Coactivations is being initialized at `init_coactivation_tracking` in SparseModule